### DISCUSSION TITLE
FIPS mode on strongswan 5.9.2

### DISCUSSION THREAD

**Discussion by gr4yg00s3** [strongswan/strongswan#515] [Q&A] [answered]
Hello, 
I've been using FIPS mode on strongswan 5.6.3 (by enabling it in /etc/stongswan.d/charon/openssl.conf), where as expected strongswan does not support md5 for authentication irrespective of the phase(1/2). All I did was I bumped strongswan to 5.9.2 and md5 now seems to be supported in phase 1 (it still is appropriately blocked for phase 2). Is this expected? If yes, is this documented somewhere? What could I do if I wanted FIPS mode to work as before, that is no matter the phase of IKE negotiation, md5 should not be supported?

**Comment by tobiasbrunner** [comment, marked as answer]
> All I did was I bumped strongswan to 5.9.2 and md5 now seems to be supported in phase 1 (it still is appropriately blocked for phase 2).

How exactly did you test this?

**Reply by gr4yg00s3** [reply]
I have a test for checking if FIPS is being enforced. So I have a test where I use MD5 for phase 1 and it is expected to fail when FIPS is enabled. It does appropriately fail in 5.6.3 (and even 5.8.1), but in 5.8.4 and 5.9.2 phase 1 MD5 seems to be able to go ahead with the IKE handshake.

**Reply by gr4yg00s3** [reply]
To test if it blocks MD5 for phase 2, I use SHA256 for phase 1 and check if phase 2 fails with MD5. It fails as expected for all versions I've tested (5.6.3, 5.8.1, 5.8.4, 5.9.2)

**Reply by tobiasbrunner** [reply]
Since 5.8.2, the _openssl_ plugin fails directly when constructing an HMAC instance if the underlying hash function is not available. This allows falling back to other implementations of the requested algorithm.  Before, it always constructed the instance successfully but later failed when it was used. So make sure you don't have some other plugin loaded that provides HMAC-MD5 (e.g. the _md5_ and _hmac_ plugins).

**Reply by gr4yg00s3** [reply]
I see. Shouldn't that also be the case for Phase 2 as well? I ask because phase 2 appropriately blocks MD5

**Reply by tobiasbrunner** [reply]
No that's completely independent as the OS kernel handles IPsec SAs (see [IntroductionTostrongSwan](https://wiki.strongswan.org/projects/strongswan/wiki/IntroductionTostrongSwan)).

**Reply by gr4yg00s3** [reply]
Oh right, of course! My mistake. FIPS mode is enabled in the kernel and hence blocks MD5 as expected. Thank you for your help with this. I will try unloading the md5 and hmac plugins and test this.

**Reply by gr4yg00s3** [reply]
Hello, I wanted to verify one more thing wrt this. What you said worked for MD5, I disbaled the md5 plugin (I let the hmac plug in stay though) and when I enable FIPS for the openssl plugin it fails to connect when MD5 is configured. Another behavior related to this is before 5.8.1 a PSK of length short than 14 characters fails to form a connection if FIPS is enabled in the openssl plugin. However given the new behavior it seems to fall back to a plugin that will let the connection have a PSK less than 14 characters. I did disable the additional plugins like sha1 and sha2, in hope that those are the plugins being used. However it would still work. But when I disable hmac plugin the IKE connection fails regardless of the length of the PSK (or algorithm). A question I had was the openssl plugin provides for everything as per this https://wiki.strongswan.org/projects/strongswan/wiki/Pluginlist so disabling the hmac plugin should have worked right? Am I missing something here?

**Reply by tobiasbrunner** [reply]
> But when I disable hmac plugin the IKE connection fails regardless of the length of the PSK.

What exactly is the error?

> so disabling the hmac plugin should have worked right? 

Unless your version of OpenSSL does not provide the HMAC API (i.e. if `OPENSSL_NO_HMAC` is defined).

You can check what algorithms the loaded plugins provide via `ipsec listalgs` or `swanctl --list-algs`.

**Reply by gr4yg00s3** [reply]
The exact error for 5.8.2 and after, on disabling the fall back plugins is (irrespective of the length of the PSK):
`PSEUDO_RANDOM_FUNCTION PRF_HMAC_SHA2_256 not supported!`
`key derivation for pre-shared key failed`

The exact error for 5.8.1 and before when PSK length < 14 characters in FIPS mode is:
`key derivation for pre-shared key failed`

The list of algorithms indicates that I should not be facing said problem? since the blow is the output
```
encryption:
AES_CBC[aes]
  AES_ECB[aes]
  3DES_CBC[des]
  DES_CBC[des]
  DES_ECB[des]
  CAMELLIA_CBC[openssl]
  CAST_CBC[openssl]
  BLOWFISH_CBC[openssl]
  NULL[openssl]
  AES_CTR[ctr]
  CAMELLIA_CTR[ctr]
integrity:
  HMAC_MD5_96[openssl]
  HMAC_MD5_128[openssl]
  HMAC_SHA1_96[openssl]
  HMAC_SHA1_128[openssl]
  HMAC_SHA1_160[openssl]
  HMAC_SHA2_256_128[openssl]
  HMAC_SHA2_256_256[openssl]
  HMAC_SHA2_384_192[openssl]
  HMAC_SHA2_384_384[openssl]
  HMAC_SHA2_512_256[openssl]
  HMAC_SHA2_512_512[openssl]
aead:
  AES_GCM_16[openssl]
  AES_GCM_12[openssl]
  AES_GCM_8[openssl]
hasher:
  HASH_SHA1[openssl]
  HASH_MD4[openssl]
  HASH_MD5[openssl]
  HASH_SHA2_224[openssl]
  HASH_SHA2_256[openssl]
  HASH_SHA2_384[openssl]
  HASH_SHA2_512[openssl]
prf:
  PRF_KEYED_SHA1[openssl]
  PRF_HMAC_MD5[openssl]
  PRF_HMAC_SHA1[openssl]
  PRF_HMAC_SHA2_256[openssl]
  PRF_HMAC_SHA2_384[openssl]
  PRF_HMAC_SHA2_512[openssl]
xof:
drbg:
dh:
  ECP_256[openssl]
  ECP_384[openssl]
  ECP_521[openssl]
  ECP_224[openssl]
  ECP_192[openssl]
  ECP_256_BP[openssl]
  ECP_384_BP[openssl]
  ECP_512_BP[openssl]
  ECP_224_BP[openssl]
  MODP_3072[openssl]
  MODP_4096[openssl]
  MODP_6144[openssl]
  MODP_8192[openssl]
  MODP_2048[openssl]
  MODP_2048_224[openssl]
  MODP_2048_256[openssl]
  MODP_1536[openssl]
  MODP_1024[openssl]
  MODP_1024_160[openssl]
  MODP_768[openssl]
  MODP_CUSTOM[openssl]
rng:
  RNG_WEAK[openssl]
  RNG_STRONG[random]
  RNG_TRUE[random]
nonce-gen:
  NONCE_GEN[nonce]
```
The list of plugins via `swanctl --stats` is:
```
uptime: 4 hours, since Aug 13 13:35:56 2021
worker threads: 16 total, 11 idle, working: 4/0/1/0
job queues: 0/0/0/0
jobs scheduled: 0
IKE_SAs: 0 total, 0 half-open
loaded plugins: charon aes des random nonce openssl ctr attr kernel-netlink socket-default vici
```

**Reply by tobiasbrunner** [reply]
The _openssl_ plugin announces support for `PRF_HMAC_SHA2_256`. However, creating an instance fails. Since 5.8.2, to ensure HMAC implementations are actually functional, `HMAC_Init_ex()` is called from the constructor by setting a key. However, that key is the empty string. So this consistently fails due to the length limit that's apparently enforced in FIPS-mode (I wasn't aware of that, is that documented anywhere or configurable?).

Strangely, that patch was done because someone used OpenSSL in FIPS mode and instantiating HMAC-MD5 worked only to fail later when actually using it, so it was not possible to fallback to the _hmac_ plugin during construction (the use case there was RADIUS, which requires HMAC-MD5).

I guess we could use a longer key for the initialization, I pushed such a change to the _515-openssl-hmac-fips_ branch.

**Reply by gr4yg00s3** [reply]
Ah that's interesting. I see, since the key is of length 0 the creation of `PRF_HMAC_SHA2_256` fails even though it is supported. 
I think I remember reading about the length limit in one of the FIPS security policy documents. Thanks for pushing a change for this. I'll try and apply the patch and see if it works okay.

**Reply by tobiasbrunner** [reply]
OpenSSL's [SecurityPolicy-2.0.16.pdf](https://www.openssl.org/docs/fips/SecurityPolicy-2.0.16.pdf) does not say anything about that. But [Ubuntu's security policy for OpenSSL in 18.04](https://csrc.nist.gov/projects/cryptographic-module-validation-program/Certificate/3622) does (link to the PDF is at the bottom in the Related Files section). And it references [SP 800-131A](https://csrc.nist.gov/publications/detail/sp/800-131a/rev-2/final), which is apparently where this restriction comes from (see section 10).

**Reply by gr4yg00s3** [reply]
This patch seems to work, I'm testing it a bit more. Thank you!

**Reply by gr4yg00s3** [reply]
The change you provided works as expected. Thank you so much for your help with this.

**Reply by tobiasbrunner** [reply]
OK, great, thanks for testing. Fixed via #557.
